### ISSUE TITLE
Difference in PETS implementation from the original TF version.

### ISSUE THREAD

**Issue by ethanluoyc** [facebookresearch/mbrl-lib#99]
This follows from the conversation in #98. I have noticed some discrepancy between the TF and mbrl-lib implementation of PETS.

## Difference in normalization.

https://github.com/kchua/handful-of-trials/blob/master/dmbrl/modeling/utils/TensorStandardScaler.py#L45

In the original version, the normalization is guarded against observation dimensions with small stddev by setting the dimensions with small stddev to 1. This prevents the normalized inputs from exploding when the stddev is small. This happens in environments such as Reacher or Pusher where some observation dimensions consist of _goals_. In that situation, it seems that the goal is never changing during an episode and the stddev will be 0. Hence setting the small stddev to be 1.0 would be helpful in that case.

Another very subtle thing happening in the above code is that the normalization is performed with NumPy instead of in TF, and I think the inputs here are in float64. In that case, the stddev computation is more accurate than those in float32, so the threshold 1e-12 is sensible. Using PyTorch to perform normalization, for example, would require changes to the threshold. I think some values like 1e-5 would be more appropriate in that case (not backed up by any numerical analysis).

## Difference in activation function
The original implementation uses the `swish` activation function whereas in mbrl-lib we use `silu`. I am confused about the choice of silu in mbrl-lib and would love to know more about the difference in empirical performance. 

## Difference in CEM stopping criteria
In the TF implementation, the CEM optimizer uses an additional termination criterion on the variance:
https://github.com/kchua/handful-of-trials/blob/77fd8802cc30b7683f0227c90527b5414c0df34c/dmbrl/misc/optimizers/cem.py#L71
I doubt that criterion is ever satisfied during training but I am mentioning this here for completeness.

## Difference in optimizer weight decay
The original TF implementation uses a carefully selected set of weight decays for different layers of the dynamics model whereas the decay in mbrl-lib is the same for all layers. However, the original implementation does not add weight decays on the biases. See

https://github.com/kchua/handful-of-trials/blob/master/dmbrl/modeling/layers/FC.py#L219

In PyTorch, the default Adam will add weight decay on all parameters. That also means that they are added to the max_logvar and min_logvar whereas in the TF version the only regularization on the max/min-logvars is through the var_loss.

Maybe a side note, have the authors tried using AdamW instead of Adam for the weight decays? I recently learned that naive weight decay in Adam does not behave as you may expect. See https://arxiv.org/abs/1711.05101

## Difference in optimizer parameters
<strike>The default epsilon in TensorFlow's Adam is 1e-7, https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam</strike>
Scratch this, they are 1e-8 in TF 1 https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/keras/optimizers/Adam.

Anyway, I am mentioning these here after a thorough look at both mbrl-lib and TF PETS to debug my own JAX implementation. Turns out my mistake was in the MPC code. I hope these notes are useful since the author mentions that currently, the current implementation does not get good performance on Half-Cheetah. Maybe it's because of one of these details, if not, fingers crossed the difference can be spotted by someone else :D

**Comment by luisenp** [comment]
Thanks for these suggestions, @ethanluoyc :) I'm particularly curious to check the effect of double normalization and maybe try again with the per-layer weight decay (I tried at some point and didn't help, but I was trying many things at that point). Depending on how that goes I may also try adding the extra stopping condition for CEM. Thanks!

BTW, as far as I know, `silu` activation is just another name for the `swish` activation ([see](https://pytorch.org/docs/stable/generated/torch.nn.SiLU.html), for example).

**Comment by ethanluoyc** [comment]
I was curious about adding layer norms to do the normalization instead of normalizing the inputs. This is in the same spirit as the networks used in model-free agents like D4PG. e.g., https://github.com/deepmind/acme/blob/master/acme/tf/networks/continuous.py#L37
If that works then there is no special need for a separate normalizer.

**Comment by luisenp** [comment]
This is worth a shot as well :)

**Comment by luisenp** [comment]
Hi @ethanluoyc. I have done some testing, and the double normalization does help a little bit, although we still can't match the original results on HalfCheetah; I am planning to merge this change anyway, and the relevant PR is #111, if you are curious. 

I also tried again the per-layer weight decay for the linear layers, but it actually seems to make things worse. Would you mind if I send you a small PR to your fork and taking a look there? Having an extra pair of eyes would be helpful :)

**Comment by ethanluoyc** [comment]
No problem at all! I can certainly take a look at it to see if I can spot something mysterious!

**Comment by luisenp** [comment]
Thanks a lot! Created #112 for this.

**Comment by luisenp** [comment]
I @ethanluoyc. I tried the weight decay per layer (as shown in the other PR) and also using AdamW, but didn't see improvements; thus I think it's a good time to close this issue. I appreciate the time you took to write this, and we did get some improvements from the double normalization. Thanks again for all the valuable suggestions! :)

**Comment by ethanluoyc** [comment]
Glad to know that AdamW does not make a difference. I don’t have enough compute to try out these things and I am really happy to learn from your experimental results!
